torchvision/models/video/mvit.py

```diff
@@ -106,28 +108,120 @@ def forward(self, x: torch.Tensor, thw: Tuple[int, int, int]) -> Tuple[torch.Ten
+    if embedding.shape[0] == d:
+        return embedding
+
+    return (
+        nn.functional.interpolate(
+            embedding.permute(1, 0).unsqueeze(0),
+            size=d,
+            mode="linear",
+        )
+        .squeeze(0)
+        .permute(1, 0)
+    )
+
+
+def _add_rel_pos(
+    attn: torch.Tensor,
+    q: torch.Tensor,
+    q_thw: Tuple[int, int, int],
+    k_thw: Tuple[int, int, int],
+    rel_pos_h: torch.Tensor,
+    rel_pos_w: torch.Tensor,
+    rel_pos_t: torch.Tensor,
+) -> torch.Tensor:
+    # Modified code from: https://github.com/facebookresearch/SlowFast/commit/1aebd71a2efad823d52b827a3deaf15a56cf4932
+    q_t, q_h, q_w = q_thw
+    k_t, k_h, k_w = k_thw
+    dh = int(2 * max(q_h, k_h) - 1)
+    dw = int(2 * max(q_w, k_w) - 1)
+    dt = int(2 * max(q_t, k_t) - 1)
+
+    # Scale up rel pos if shapes for q and k are different.
+    q_h_ratio = max(k_h / q_h, 1.0)
+    k_h_ratio = max(q_h / k_h, 1.0)
+    dist_h = torch.arange(q_h)[:, None] * q_h_ratio - (torch.arange(k_h)[None, :] + (1.0 - k_h)) * k_h_ratio
+    q_w_ratio = max(k_w / q_w, 1.0)
+    k_w_ratio = max(q_w / k_w, 1.0)
+    dist_w = torch.arange(q_w)[:, None] * q_w_ratio - (torch.arange(k_w)[None, :] + (1.0 - k_w)) * k_w_ratio
+    q_t_ratio = max(k_t / q_t, 1.0)
+    k_t_ratio = max(q_t / k_t, 1.0)
+    dist_t = torch.arange(q_t)[:, None] * q_t_ratio - (torch.arange(k_t)[None, :] + (1.0 - k_t)) * k_t_ratio
+
+    # Intepolate rel pos if needed.
+    rel_pos_h = _interpolate(rel_pos_h, dh)
+    rel_pos_w = _interpolate(rel_pos_w, dw)
+    rel_pos_t = _interpolate(rel_pos_t, dt)
+    Rh = rel_pos_h[dist_h.long()]
+    Rw = rel_pos_w[dist_w.long()]
+    Rt = rel_pos_t[dist_t.long()]
+
+    B, n_head, _, dim = q.shape
+
+    r_q = q[:, :, 1:].reshape(B, n_head, q_t, q_h, q_w, dim)
+    rel_h_q = torch.einsum("bythwc,hkc->bythwk", r_q, Rh)  # [B, H, q_t, qh, qw, k_h]
+    rel_w_q = torch.einsum("bythwc,wkc->bythwk", r_q, Rw)  # [B, H, q_t, qh, qw, k_w]
+    # [B, H, q_t, q_h, q_w, dim] -> [q_t, B, H, q_h, q_w, dim] -> [q_t, B*H*q_h*q_w, dim]
+    r_q = r_q.permute(2, 0, 1, 3, 4, 5).reshape(q_t, B * n_head * q_h * q_w, dim)
+    # [q_t, B*H*q_h*q_w, dim] * [q_t, dim, k_t] = [q_t, B*H*q_h*q_w, k_t] -> [B*H*q_h*q_w, q_t, k_t]
+    rel_q_t = torch.matmul(r_q, Rt.transpose(1, 2)).transpose(0, 1)
+    # [B*H*q_h*q_w, q_t, k_t] -> [B, H, q_t, q_h, q_w, k_t]
+    rel_q_t = rel_q_t.view(B, n_head, q_h, q_w, q_t, k_t).permute(0, 1, 4, 2, 3, 5)
+
+    # Combine rel pos.
+    rel_pos = (
+        rel_h_q[:, :, :, :, :, None, :, None]
+        + rel_w_q[:, :, :, :, :, None, None, :]
+        + rel_q_t[:, :, :, :, :, :, None, None]
+    ).reshape(B, n_head, q_t * q_h * q_w, k_t * k_h * k_w)
+
+    # Add it to attention
+    attn[:, :, 1:, 1:] += rel_pos
+
+    return attn
+
+
+def _add_shortcut(x: torch.Tensor, shortcut: torch.Tensor, skip_cls: bool):
+    if skip_cls:
+        x[:, :, 1:, :] += shortcut[:, :, 1:, :]
+    else:
+        x.add_(shortcut)
+    return x
```

Comment:
@lyttonhao @haooooooqi @feichtenhofer I got one more question for you concerning how the shortcut in the attention is added. I believe there is a difference between the two reference implementations of SlowFast and PyTorch Video:
- PTV seems to [straight add](https://github.com/facebookresearch/pytorchvideo/blob/718d0a46a8ec3cfd12b823a2063ab66b006281b2/pytorchvideo/layers/attention.py#L488) q in the outcome of attn @ v. 
- Slowfast seems to [condition](https://github.com/facebookresearch/SlowFast/blob/1aebd71a2efad823d52b827a3deaf15a56cf4932/slowfast/models/attention.py#L432-L435) how this addition will happen based on the value of `has_cls_embed`. 

Both implementations support `has_cls_embed`, but each handles it differently (note that in TorchVision this value is always True, so it's omitted). My questions to you are:
- Why is there a difference between the two implementations?
- Do you think the way I handle the difference here makes sense?

Reply:
I think they are just two implementations while the SlowFast one is our original implementation. Regarding the performance difference, our ablation shows the results are very close. So I think as long as the training/testing are consistent, this should not be an issue.

The `skip_cls` implementation looks good to me!

Reply:
Thanks for letting me know! 

Do you think it's OK to populating `skip_cls` with the value of `rel_pos` makes sense or we should introduce another parameter for the Attention layer? 
https://github.com/pytorch/vision/blob/be6d7e080a9b22de59089efd8268f99c19dfa5c9/torchvision/models/video/mvit.py#L318

> So I think as long as the training/testing are consistent, this should not be an issue.

Do you happen to have pre-trained weights that don't require this. If yes, I would rather use them instead and avoid the extra parameter.

Reply:
I had an offline discussion with @lyttonhao where he suggested that the `skip_cls` should be independent from the `rel_pos`. Unfortunately this difference on the implementations between SlowFast and PyTorchVideo means that we will need an extra config to control for it. Based on his feedback, I introduced an extra parameter called `residual_cls`. If True, the addition in the residual connection includes the class embeddings. If False, the class embeddings are excluded in the addition. This parameter is exposed in the constructor of MViT.